### PR TITLE
Add amoCRM provider

### PR DESCRIPTION
**Pull request by denis-ivanov**
Add amoCRM provider

- Add amoCRM provider and unit test
- Update README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,6 +109,7 @@ If a provider you're looking for does not exist, consider making a PR to add one
 
 | Provider | Stable | Nightly | Documentation |
 |:-:|:-:|:-:|:-:|
+| amoCRM | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.AmoCrm?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.AmoCrm/ "Download AspNet.Security.OAuth.AmoCrm from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.AmoCrm?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.AmoCrm "Download AspNet.Security.OAuth.AmoCrm from MyGet.org") | [Documentation](https://www.amocrm.com/developers/content/oauth/step-by-step/ "amoCRM developer documentation") |
```
Comment (martincostello) [inline README.md]: _amo_ comes after _ama_

File: src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationDefaults.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        /// <summary>
+        /// Default value for the <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "https://www.amocrm.ru/oauth";
+
+        /// <summary>
+        /// Default value for the <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpoint = "https://{0}.amocrm.ru/oauth2/access_token";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationDefaults.cs]: Add `Format` suffix as this isn't a well-formed URI in of itself due to the format parameter.

See the Shopify provider for an example:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/985f785d7497f83f35f25eee8ea2de6fad7d2fc9/src/AspNet.Security.OAuth.Shopify/ShopifyAuthenticationDefaults.cs#L40

Alternatively you can use the pattern used by the Okta provider that uses paths and `IPostConfigureOptions<T>`:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/985f785d7497f83f35f25eee8ea2de6fad7d2fc9/src/AspNet.Security.OAuth.Okta/OktaAuthenticationDefaults.cs#L45

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/985f785d7497f83f35f25eee8ea2de6fad7d2fc9/src/AspNet.Security.OAuth.Okta/OktaPostConfigureOptions.cs#L19-L31

File: src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            AuthorizationEndpoint = AmoCrmAuthenticationDefaults.AuthorizationEndpoint;
+
+            ClaimActions.MapJsonKey(ClaimTypes.Email, "email");
+            ClaimActions.MapJsonKey(ClaimTypes.Name, "name");
+            ClaimActions.MapJsonKey(ClaimTypes.Surname, "last_name");
+            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "id");
+        }
+
+        /// <summary>
+        /// amoCRM account name.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs]:
```suggestion
        /// Gets or sets the amoCRM account name.
```

File: src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        /// <exception cref="ArgumentException">Thrown when <paramref name="value"/> is null or empty string.</exception>
+        public string Account
+        {
+            get => _account;
+            set
+            {
+                if (string.IsNullOrWhiteSpace(value))
+                {
+                    throw new ArgumentException("Invalid account", nameof(value));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs]:
```suggestion
                    throw new ArgumentException("Account cannot be null or white space.", nameof(value));
```

File: test/AspNet.Security.OAuth.Providers.Tests/AmoCrm/AmoCrmTests.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                options.Account = "example";
+                ConfigureDefaults(builder, options);
+            });
+        }
+
+        [Theory]
+        [InlineData(ClaimTypes.NameIdentifier, "100500")]
+        [InlineData(ClaimTypes.Name, "John")]
+        [InlineData(ClaimTypes.Email, "john@john-smith.local")]
+        public async Task Can_Sign_In_Using_AmoCrm(string claimType, string claimValue)
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/AmoCrm/AmoCrmTests.cs:33]: Please add a test case for `ClaimTypes.Surname`.

File: README.md (hunk 1 above)
Comment (denis-ivanov) [reply README.md]: @martincostello, done!

File: src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationDefaults.cs (hunk 2 above)
Comment (denis-ivanov) [reply src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationDefaults.cs]: @martincostello, done!

File: src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs (hunk 3 above)
Comment (denis-ivanov) [reply src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs]: @martincostello, done!

File: src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs (hunk 4 above)
Comment (denis-ivanov) [reply src/AspNet.Security.OAuth.AmoCrm/AmoCrmAuthenticationOptions.cs]: @martincostello, done!

File: test/AspNet.Security.OAuth.Providers.Tests/AmoCrm/AmoCrmTests.cs (hunk 5 above)
Comment (denis-ivanov) [reply test/AspNet.Security.OAuth.Providers.Tests/AmoCrm/AmoCrmTests.cs:33]: @martincostello, done!

**Comment by martincostello** [comment]
Thanks @denis-ivanov - this looks good!

Once we've made a decision on our branch strategy for 3.1 and 5.0, I'll look at merging this PR and sorting out a release for it.

**Comment by denis-ivanov** [comment]
@martincostello, ok, thank you!

**Comment by martincostello** [comment]
Hi @denis-ivanov - could you also add some documentation about the extra parameter like we have for the other providers please? https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/tree/dev/docs

I'd forgotten about this when I first reviewed the PR 😅

File: docs/amocrm.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        {
+            options.Account = "my-account";
+            options.ClientId = "client-id";
+            options.ClientSecret = "client-secret";
+        });
+```
+
+## Required Additional Settings
+
+_None._
```
Comment (martincostello) [inline docs/amocrm.md]: `Account` is required, no?

**Comment by martincostello** [comment]
Thanks for you contribution! 🎈

The provider will be released to NuGet.org within the next week or so once I've got some admin tasks on the repository completed.

In the meantime, you can consume a prerelease version of the provider from our MyGet feed once the CI is completed: https://www.myget.org/gallery/aspnet-contrib

**Comment by martincostello** [comment]
Hi @denis-ivanov - I've just noticed that you have already published a package to NuGet.org with this ID. I didn't think to check this before, my bad I usually check.

In order for us to publish your contribution as part of our release process you will need to transfer ownership of the package in NuGet.org from yourself to the `aspnet-contrib` organisation. Otherwise when we try to publish the package we'll get an HTTP 403 error.

I'm afraid we can't accept **co**-ownership as we need to guarantee that any changes to the packages come from from this repository and our NuGet organisation, and not elsewhere (which would be possible via your NuGet API keys otherwise).

It's up to you whether to agree to do this, but otherwise the only options available to us are:
1. Revert this PR.
2. Rename the provider to not have the same name as the one you have already published (if we can come up with a name that doesn't clash but still makes sense).

As merging this has created a block on us releasing packages to NuGet.org, we need to do one of the two.

If we get no response from you within the next 7 days, we'll go with option 1 and remove the provider from the repository.

**Comment by denis-ivanov** [comment]
@martincostello, I'm ok with transfering ownership. I've sent you invite.

**Comment by martincostello** [comment]
Thanks Denis. Sorry for not catching that before the merge.

We had this happen once before and it created a bit of...drama.

@kevinchalet - could you handle the transfer process in NuGet when you get a moment please?

**Comment by kevinchalet** [comment]
@denis-ivanov thanks! 👏🏻 

@martincostello done :smiley:

**Comment by martincostello** [comment]
Thanks again for your contribution @denis-ivanov - the provider is now available from NuGet.org as versions 3.1.7 and 5.0.1.

**Pull request merged**